### PR TITLE
Http request counter

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: #1737 

- Move StartTry to private.
- Adding GetRetryCount().

### CODE REVIEW

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 1]
```diff
@@ -588,9 +600,13 @@ namespace Azure { namespace Core { namespace Http {
...
-    // Expected to be called by a Retry policy to reset all headers set after this function was
-    // previously called
-    void StartTry();
+
+    /**
+     * @brief Get the number of retries intents for the request.
+     *
+     * @return uint
+     */
+    uint GetRetryCount() const { return m_retryCount; }
```
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:609]:
```suggestion
    int GetRetryCount() const { return m_retryCount; }
```

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 2]
```diff
@@ -462,6 +468,12 @@ namespace Azure { namespace Core { namespace Http {
     // read and upload chunks of data from the payload body stream. If it is not set, the transport
     // adapter will decide chunk size.
     int64_t m_uploadChunkSize = 0;
+    // The count gets incremented on every intent to send the request.
+    uint m_retryCount = 0;
```
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:472]:
```suggestion
    int m_retryCount = 0;
```

**Review by RickWinter — CHANGES_REQUESTED** [review]
The original issue this fixes needs more details into the problem we are fixing.

**Comment by vhvb1989** [comment]
> The original issue this fixes needs more details into the problem we are fixing.

Other languages uses a `RequestContext` object to share one state across the pipeline.
This feature is require to implement a policy that reacts to some of the output of a previous policy. For example, the `StorageSwitchServerPolicy` needs to know if the request is on retry mode or not in order to switch to a second server.

I consider adding the entire RequestContext object as a map<string, string> to allow any state to be recorded, but this becomes more expensive for what we currently need right now ( just one int value)

**Comment by ahsonkhan** [comment]
> I consider adding the entire RequestContext object as a map<string, string> to allow any state to be recorded, but this becomes more expensive for what we currently need right now ( just one int value)

Are we sure we won't need to add more context info across different retries in the future? Having a struct wrapping the int allows us to be more extensible in the future without having to make breaking changes.

**Comment by vhvb1989** [comment]
> Are we sure we won't need to add more context info across different retries in the future? 

Just for now. God knows about the future hahahha..  But, if the future comes, the approach would be to keep adding the state as private fields in the request.  I think this is more efficient for C++ than a generic `map<string, T>`

**Comment by vhvb1989** [comment]
Use Azure Core context

**Comment by vhvb1989** [comment]
@Jinming-Hu  After reviewing this solution with Azure Core team, we decided to use the Context to keep the retry number information.
Please use `Azure::Core::Http::RetryProlicy::GetRetryNumber(Context)` to know about the retry number from the context.

The retry policy will update the value on each retry

**Comment by Jinming-Hu** [comment]
> @Jinming-Hu After reviewing this solution with Azure Core team, we decided to use the Context to keep the retry number information.
> Please use `Azure::Core::Http::RetryProlicy::GetRetryNumber(Context)` to know about the retry number from the context.
> 
> The retry policy will update the value on each retry

I don't see any problem in this way. So it works for me. We may need to merge other PRs first.

File: sdk/core/azure-core/src/http/retry_policy.cpp [hunk 3]
```diff
@@ -122,15 +122,45 @@ bool ShouldRetryOnResponse(
 
   return true;
 }
+
+static constexpr char RetryKey[] = "azureSdkRetryPolicyCounter";
```
Comment (RickWinter) [inline sdk/core/azure-core/src/http/retry_policy.cpp:126]:
```suggestion
static constexpr char RetryKey[] = "AzureSdkRetryPolicyCounter";
```

File: sdk/core/azure-core/src/http/retry_policy.cpp [hunk 4]
```diff
@@ -122,15 +122,45 @@ bool ShouldRetryOnResponse(
...
+
+/**
+ * @brief Creates a new #Context node from \p parent with the information about the retrying while
+ * sending an Http request.
+ *
+ * @param parent The parent context for the new created.
+ * @return Context with information about retry counter.
+ */
+Context inline GetRetryContext(Context const& parent)
```
Comment (RickWinter) [inline sdk/core/azure-core/src/http/retry_policy.cpp:135]:
```suggestion
Context inline CreateRetryContext(Context const& parent)
```

File: sdk/core/azure-core/src/http/retry_policy.cpp [hunk 5]
```diff
@@ -122,15 +122,45 @@ bool ShouldRetryOnResponse(
...
+  return parent.WithValue(RetryKey, retryCount);
+}
 } // namespace
 
+int Azure::Core::Http::RetryPolicy::GetRetryNumber(Context const& context)
+{
+  if (!context.HasKey(RetryKey))
+  {
+    // Not in retry mode.
+    return 0;
```
Comment (RickWinter) [inline sdk/core/azure-core/src/http/retry_policy.cpp:152]: Document that 0 implies no RetyPolicy is present.

Comment (RickWinter) [reply sdk/core/azure-core/src/http/retry_policy.cpp:152]: -1 is a better sentinel

**Review by RickWinter — APPROVED** [review]

**Pull request merged**